examples/jvm-interop/bridge.c

```diff
@@ -0,0 +1,387 @@
+{
+    uint8_t *bytes = (uint8_t *)&str;
+    uint8_t last_byte = bytes[sizeof(str) - 1];
+    uint8_t last_byte_xored = last_byte ^ 0b10000000;
+    size_t small_len = (size_t)(last_byte_xored);
+    size_t big_len = str.len;
+
+    // Avoid branch misprediction costs by always
+    // determining both small_len and big_len,
+    // so this compiles to a cmov instruction.
+    if (is_small_str(str))
+    {
+        return small_len;
+    }
+    else
+    {
+        return big_len;
+    }
+}
+
+
+extern void roc__programForHost_1__InterpolateString_caller(struct RocStr *name, char *closure_data, struct RocStr *ret);
+
+extern void roc__programForHost_1__MulArrByScalar_caller(struct RocBytesI32 *arr, int32_t *scalar, char *closure_data, struct RocBytesI32 *ret);
+
+extern void roc__programForHost_1__Factorial_caller(int64_t *scalar, char *closure_data, int64_t *ret);
+
+
+JNIEXPORT jstring JNICALL Java_javaSource_Demo_sayHello
+   (JNIEnv *env, jobject thisObj, jstring name)
+{
+    const char *jnameChars = (*env)->GetStringUTFChars(env, name, 0);
+    // we copy just in case the jvm would try to reclaim that mem
+    uint8_t *cnameChars = (uint8_t *)strdup(jnameChars);
+    size_t nameLength = (size_t) (*env)->GetStringLength(env, name);
+    (*env)->ReleaseStringUTFChars(env, name, jnameChars);
+
+
+    struct RocStr rocName = init_rocstr(cnameChars, nameLength);
+    struct RocStr ret = {0};
+
+    // Call the Roc function to populate `ret`'s bytes.
+    roc__programForHost_1__InterpolateString_caller(&rocName, 0, &ret);
+    jbyte *bytes = (jbyte*)(is_small_str(ret) ? (uint8_t*)&ret : ret.bytes);
+
+    // java being java making this a lot harder than it needs to be
+    // https://stackoverflow.com/questions/32205446/getting-true-utf-8-characters-in-java-jni
+    // https://docs.oracle.com/javase/1.5.0/docs/guide/jni/spec/types.html#wp16542
+    // but as i refuse converting those manually to their correct form, we just let the jvm handle the conversion
+    // by first making a java byte array then converting the byte array to our final jstring
+    jbyteArray byteArray = (*env)->NewByteArray(env, ret.len);
+    (*env)->SetByteArrayRegion(env, byteArray, 0, ret.len, bytes);
+    /* jbyte* bytes = (*env)->GetByteArrayElements(env, byteArray, 0); */
+
+    jstring charsetName = (*env)->NewStringUTF(env, "UTF-8");
+    jclass stringClass = (*env)->FindClass(env, "java/lang/String");
+    // https://docs.oracle.com/javase/7/docs/jdk/api/jpda/jdi/com/sun/jdi/doc-files/signature.html
+    jmethodID stringConstructor = (*env)->GetMethodID(env, stringClass, "<init>", "([BLjava/lang/String;)V");
+    jstring result = (*env)->NewObject(env, stringClass, stringConstructor, byteArray, charsetName);
+
+    // cleanup
+    if (is_seamless_str_slice(ret)) {
+      decref((void *)(ret.capacity << 1), alignof(uint8_t *));
+    }
+    else {
+      decref(ret.bytes, alignof(uint8_t *));
+    }
+    decref((void *)&rocName, alignof(uint8_t *));
```

Comment:
No need to free rocName. Roc should handle it.

Reply:
done